## 3. Библиотека requests

In [3]:
import requests

Как только библиотека установлена, импортируем её и отправим наш первый запрос к ресурсу Курсы валют ЦБ РФ в XML и JSON. Используем метод get() из библиотеки requests, передав ему соответствующий URL — https://www.cbr-xml-daily.ru/daily_json.js:>

In [11]:
url = "https://www.cbr-xml-daily.ru/daily_json.js"  # Определяем значение URL страницы для запроса
response = requests.get(
    url
)  # Делаем GET-запрос к ресурсу и результат ответа сохраняем в переменной response

Проверим ответ сервера — содержимое переменной response:

In [ ]:
print(response)  # Выводим значение response на экран как объект

<Response [200]>


Мы получили объект ответа Response, который содержит всю нужную нам информацию. По умолчанию в квадратных скобках на экран выводится код статуса ответа. В данном случае он равен 200 — то есть запрос был корректным и сервер отдал нам нужную информацию. Значение кода статуса 404 означало бы, что страница по указанному адресу не найдена, а значение 403 — что синтаксис GET - запроса неверный.

Код ответа в виде числовой переменной можно получить с помощью метода status_code:

In [ ]:
print(response.status_code)  # Выводим числовое значение response на экран

200


### Задание 3.1

Вы уже импортировали модуль requests в ваш код. Напишите строку кода, при выполнении которой будет сделан GET-запрос к ресурсу https://www.cbr-xml-daily.ru/daily.xml, а результат ответа будет сохранён в переменной response:

In [8]:
response = requests.get("https://www.cbr-xml-daily.ru/daily.xml")
response

<Response [200]>

Мы сделали запрос и получили корректный ответ (код статуса — 200). Дальнейшую работу производим с результатом запроса к ресурсу Курсы валют ЦБ РФ вXML и JSON.

### Как получить доступ ко всей информации, которую содержит ответ?
Текст ответа хранится в атрибуте text. Выведем значение атрибута на экран и посмотрим на его содержимое:

In [12]:
print(response.text)  # Выводим содержимое атрибута text переменной response на экран

{
    "Date": "2024-11-26T11:30:00+03:00",
    "PreviousDate": "2024-11-23T11:30:00+03:00",
    "PreviousURL": "\/\/www.cbr-xml-daily.ru\/archive\/2024\/11\/23\/daily_json.js",
    "Timestamp": "2024-11-25T20:00:00+03:00",
    "Valute": {
        "AUD": {
            "ID": "R01010",
            "NumCode": "036",
            "CharCode": "AUD",
            "Nominal": 1,
            "Name": "Австралийский доллар",
            "Value": 67.7235,
            "Previous": 66.7155
        },
        "AZN": {
            "ID": "R01020A",
            "NumCode": "944",
            "CharCode": "AZN",
            "Nominal": 1,
            "Name": "Азербайджанский манат",
            "Value": 61.0534,
            "Previous": 60.3389
        },
        "GBP": {
            "ID": "R01035",
            "NumCode": "826",
            "CharCode": "GBP",
            "Nominal": 1,
            "Name": "Фунт стерлингов Соединенного королевства",
            "Value": 130.0187,
            "Previous": 129.4305
 

Для того чтобы удобно было работать с полученной информацией, нам необходимо преобразовать строку в словарь. В объект ответа Response  из библиотеки requests уже встроен метод json().

Импортируем функцию pprint(), применим к полученному ответу метод json() и выведем полученный результат на экран:

In [13]:
from pprint import pprint  # Импортируем функцию pprint()

currencies = response.json()  # Применяем метод json()
pprint(currencies)  # Выводим результат на экран)

{'Date': '2024-11-26T11:30:00+03:00',
 'PreviousDate': '2024-11-23T11:30:00+03:00',
 'PreviousURL': '//www.cbr-xml-daily.ru/archive/2024/11/23/daily_json.js',
 'Timestamp': '2024-11-25T20:00:00+03:00',
 'Valute': {'AED': {'CharCode': 'AED',
                    'ID': 'R01230',
                    'Name': 'Дирхам ОАЭ',
                    'Nominal': 1,
                    'NumCode': '784',
                    'Previous': 27.9309,
                    'Value': 28.2616},
            'AMD': {'CharCode': 'AMD',
                    'ID': 'R01060',
                    'Name': 'Армянских драмов',
                    'Nominal': 100,
                    'NumCode': '051',
                    'Previous': 26.3178,
                    'Value': 26.6035},
            'AUD': {'CharCode': 'AUD',
                    'ID': 'R01010',
                    'Name': 'Австралийский доллар',
                    'Nominal': 1,
                    'NumCode': '036',
                    'Previous': 66.7155,
            

Теперь данные находятся в словаре и можно легко получать необходимые значения.

Например, по ключу Valute мы можем обратиться к вложенному словарю, который содержит информацию о мировых валютах. Выведем на экран, например, информацию о евро (EUR):

In [14]:
pprint(currencies["Valute"]["EUR"])  # Выводим на экран информацию о валюте евро

{'CharCode': 'EUR',
 'ID': 'R01239',
 'Name': 'Евро',
 'Nominal': 1,
 'NumCode': '978',
 'Previous': 107.4252,
 'Value': 108.8705}


### Задание 3.2
Повторите запросы, описанные в этом юните, на своём компьютере. Что будет выведено на экран в результате выполнения следующего кода?

```python
print(currencies['Valute']['CZK']['Name'])
```

In [ ]:
print(currencies["Valute"]["CZK"]["Name"])

Чешских крон


## 4. Парсинг сайтов

### Получаем содержимое веб-страницы

Получим HTML-код интересующей нас страницы.

Для этого отправим GET-запрос с помощью библиотеки requests и метода get() и посмотрим на текст ответа на наш запрос (как мы помним, он содержится в атрибуте text):

In [16]:
url = "https://nplus1.ru/news/2021/10/11/econobel2021"  # Определяем адрес страницы
response = requests.get(url)  # Выполняем GET-запрос
print(response.text)  # Выводим содержимое атрибута text

<!DOCTYPE html>
<html prefix="og: http://ogp.me/ns#" lang="ru">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Премию Нобеля по экономике присудили за исследования экономики труда и причинно-следственных связей</title>
    <link rel="preload" href="https://staticn1.nplus1.ru/fonts/AeonikPro/AeonikPro-Regular.woff2" as="font" type="font/woff2" crossorigin />
    <link rel="preload" href="https://staticn1.nplus1.ru/fonts/Spectral/Spectral-Regular.woff" as="font" type="font/woff2" crossorigin />
  <link href="/front-build/css/main.css?id=842e64708435127ee441c2e86afffea2" rel="stylesheet">
  <link href="/front-build/css/app.css?id=92673ebe7185ebc42e29fc7294a4c65f" rel="stylesheet">
  

  <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
  <link rel="

В отличие от предыдущего примера, где ответ возвращался в JSON-формате, мы не можем так просто преобразовать HTML-код в словарь и извлечь необходимую нам информацию.

Для решения таких задач в Python существует специальная библиотека BeautifulSoup, о работе с которой мы поговорим в следующем юните.

## 5. Библиотека BeautifulSoup

Для поиска необходимых нам данных мы будем использовать библиотеку BeautifulSoup, которая позволяет по названию тегов и их атрибутов получать содержащийся в них текст.

In [17]:
from bs4 import BeautifulSoup

Ранее мы уже получили содержимое страницы с помощью GET-запроса и сохранили информацию в переменной response, теперь создадим объект BeautifulSoup с именем page, указывая в качестве параметра html.parser.

Для примера получим информацию o title (с англ. заголовок) — это строка, которая отображается на вкладке браузера:

In [ ]:
url = "https://nplus1.ru/news/2021/10/11/econobel2021"  # Определяем адрес страницы
response = requests.get(
    url
)  # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(
    response.text, "html.parser"
)  # Создаём объект BeautifulSoup, указывая html-парсер
print(page.title)  # Получаем тег title, отображающийся на вкладке браузера
print(
    page.title.text
)  # Выводим текст из полученного тега, который содержится в атрибуте text

<title>Премию Нобеля по экономике присудили за исследования экономики труда и причинно-следственных связей</title>
Премию Нобеля по экономике присудили за исследования экономики труда и причинно-следственных связей


### Извлекаем заголовок и время написания статьи

Предположим, что мы знаем, что в HTML-коде рассматриваемой нами страницы заголовок статьи заключён в тег <`h1`> … <`/h1`> (заголовок первого уровня).

Тогда мы можем получить его текст с помощью метода find() (с англ. найти) объекта BeautifulSoup, передав ему название интересующего нас тега:

In [24]:
print(
    page.find("h1").text
)  # Применяем метод find() к объекту и выводим результат на экран


            Премию Нобеля по экономике присудили за исследования экономики труда и причинно-следственных связей
          


Теперь получим данные из него с помощью метода find(), передав название нужного тега:

In [40]:
print(page.find("span", class_="relative").text)


13:04



### Задание 5.3 

Используя инструмент разработчика в браузере, определите, в каком теге лежит большой заголовок на странице в Wikipedia.

In [32]:
def wiki_header(url):
    response = requests.get(url)
    page = BeautifulSoup(response.text, "html.parser")
    return page.find("h1").text

In [ ]:
wiki_header("https://en.wikipedia.org/wiki/Operating_system")

'Operating system'

### Неуникальные теги: извлекаем текст и дату публикации статьи

Посмотрим на нашу страницу, используя инструмент разработчика, ещё раз. Можем заметить, что у искомого текста есть свой класс — n1_material text-18 

In [34]:
print(
    page.find("div", class_="n1_material text-18").text
)  # Выводим содержимое атрибута text тега div класса n1_material text-18

Премия Шведского национального банка по экономическим наукам памяти Альфреда Нобеля за 2021 год присуждена Дэвиду Карду (David Card) за его вклад в эмпирические исследования экономики рынка труда, а также Джошуа Энгристу (Joshua Angrist) и Гвидо Имбенсу (Guido Imbens) за их вклад в методологию анализа причинно-следственных связей. Прямая трансляция церемонии объявления лауреатов шла на официальном сайте Нобелевской премии.


В данном случае происходит поиск точного строкового значения class атрибута, т. е. выполнение строк кода

In [ ]:
print(page.find("div", class_="n1_material").text)

Премия Шведского национального банка по экономическим наукам памяти Альфреда Нобеля за 2021 год присуждена Дэвиду Карду (David Card) за его вклад в эмпирические исследования экономики рынка труда, а также Джошуа Энгристу (Joshua Angrist) и Гвидо Имбенсу (Guido Imbens) за их вклад в методологию анализа причинно-следственных связей. Прямая трансляция церемонии объявления лауреатов шла на официальном сайте Нобелевской премии.


In [ ]:
print(page.find("div", class_="n1_material text-18").text)

Премия Шведского национального банка по экономическим наукам памяти Альфреда Нобеля за 2021 год присуждена Дэвиду Карду (David Card) за его вклад в эмпирические исследования экономики рынка труда, а также Джошуа Энгристу (Joshua Angrist) и Гвидо Имбенсу (Guido Imbens) за их вклад в методологию анализа причинно-следственных связей. Прямая трансляция церемонии объявления лауреатов шла на официальном сайте Нобелевской премии.


даст одинаковый результат.

При выполнении строки кода

In [37]:
print(page.find("div", class_="text-18 n1_material").text)

AttributeError: 'NoneType' object has no attribute 'text'

мы получим ошибку, так как такого строкового значения в области поиска нет.

Аналогично получим информации о теге, который содержит дату написания статьи, отображаемую в левом верхнем углу страницы.

Итак, нам нужен тег `<a>` … `</a>` с классом "relative before:block before:w-px before:bg-current before:h-4 before:absolute before:left-0 group pl-2 flex inline-flex items-center". Для поиска достаточно указать в качестве класса "relative", отбросив дополнительные настройки.

Теперь получим данные из него с помощью уже известного метода find(), передав название нужного тега:

In [41]:
# Выводим на экран содержимое атрибута text тега a с классом "relative"
print(page.find("a", class_="relative").text)


11.10.21



Задача решена — мы извлекли из контента страницы заголовок статьи, опубликованной на странице, дату публикации, а также текст статьи.

### Сбор нескольких элементов: собираем все ссылки на странице

Рассмотрим ещё один сценарий: вы хотите собрать сразу несколько элементов со страницы. Например, представьте, что вы хотите получить названия всех языков программирования, упомянутых на странице в Wikipedia в статье про языки программирования.

Можно заметить, что все названия языков программирования на этой странице связаны ссылками c соответствующими статьями о них. Таким образом, нам необходимо собрать все ссылки на странице. Для ссылок в HTML предусмотрен тег `<a>` … `</a>`. Попробуем использовать find():

In [42]:
url = "https://en.wikipedia.org/wiki/List_of_programming_languages"  # Задаём адрес ресурса
response = requests.get(url)  # Делаем GET-запрос к ресурсу
page = BeautifulSoup(response.text, "html.parser")  # Создаём объект BeautifulSoup
print(page.find("a"))  # Ищем ссылку по тегу <a> и выводим её на экран

<a class="mw-jump-link" href="#bodyContent">Jump to content</a>


Мы получили только одну ссылку, хотя на странице их явно больше.

Это происходит, потому что метод find() возвращает только первый подходящий элемент. Если требуется получить больше элементов, необходимо воспользоваться методом find_all() (с англ. найти все):

In [43]:
links = page.find_all(
    "a"
)  # Ищем все ссылки на странице и сохраняем в переменной links в виде списка
print(len(links))  # Выводим количество найденных ссылок

950


Итак, на момент создания этих учебных материалов на странице содержалось 950 ссылок. Посмотрим на некоторые из них:

In [44]:
print(
    [link.text for link in links[500:510]]
)  # Выводим ссылки с 500 по 509 включительно

['MAD', 'MAD/I', 'Magik', 'Magma', 'Maple', 'MAPPER', 'MARK-IV', 'Mary', 'MATLAB', 'MASM Microsoft Assembly x86']


Не все ссылки соответствуют названиям языков программирования — страница содержит также «служебные» ссылки, такие, например, как Jump to navigation (с англ. Перейти к навигации) или Alphabetical (с англ. По алфавиту):

In [45]:
print([link.text for link in links[0:10]])  # Выводим ссылки с 1 по 9 включительно

['Jump to content', 'Main page', 'Contents', 'Current events', 'Random article', 'About Wikipedia', 'Contact us', 'Help', 'Learn to edit', 'Community portal']


## 6. Работа с API

Токен

fc83f2d0fc83f2d0fc83f2d038ffa6c50effc83fc83f2d09bc4adc5aad2da07209726b1

### Первые запросы к API

Чтобы познакомиться с работой API, мы будем получать данные для статистических отчётов произвольной группы, например данные о соотношении мужчин и женщин, статистику географии пользователей и т. п.

Сначала рассмотрим работу API на простом примере, на основе которого работают многие системы.

Сделаем наш первый запрос из браузера.

Перейдите по следующей ниже ссылке в браузере, подставив вместо слова TOKEN ваш персональный сервисный ключ доступа (токен), полученный на предыдущем шаге:

https://api.vk.com/method/users.get?user_id=1&v=5.95&access_token=fc83f2d0fc83f2d0fc83f2d038ffa6c50effc83fc83f2d09bc4adc5aad2da07209726b1

Результат:

```
{"response":[{"id":1,"first_name":"Павел","last_name":"Дуров",  
"can_access_closed":true,"is_closed":false}]}
```

Итак, мы сделали GET-запрос к API ВКонтакте, который состоит из следующих элементов:

- https://api.vk.com/method — домен и URL запроса API; обычно не меняется;
- users.get — название метода, который отдаёт определённый отчёт, в нашем случае это метод для получения информации о пользователе;
- user_id и v — параметры запроса: идентификатор пользователя, о котором хотим получить информацию (в нашем примере мы запрашиваем информацию о первом пользователе), и номер версии API;
- token — токен, который выдаётся только пользователям, имеющим право просматривать определённые данные, например показания счётчиков Яндекс.Метрики вашего проекта; на все остальные запросы без корректного токена система отвечает отказом.

Если мы обратимся к документации метода users.get, то увидим, что в ней описано множество других параметров, которые можно получить о пользователе (дата рождения, пол, родной город и другие) — словом, всё то, что мы видим на странице пользователя в интерфейсе или приложении ВКонтакте (конечно, если пользователь их указал).

Добавим к запросу дату рождения и пол (согласно документации, эти параметры надо перечислять в поле fields):

https://api.vk.com/method/users.get?user_id=1&v=5.95&fields=sex,bdate&access_token=fc83f2d0fc83f2d0fc83f2d038ffa6c50effc83fc83f2d09bc4adc5aad2da07209726b1

```
{"response":[{"id":1,"bdate":"10.10.1984","sex":2,"first_name":"Павел",  "last_name":"Дуров","can_access_closed":true,"is_closed":false}]}
```

До этого момента мы делали запросы в браузере — теперь давайте выполним запросы из кода.

### Запрос к API из кода

In [20]:
import requests

token = "fc83f2d0fc83f2d0fc83f2d038ffa6c50effc83fc83f2d09bc4adc5aad2da07209726b1"  # Указываем свой сервисный токен
url = "https://api.vk.com/method/users.get"  # Указываем адрес страницы к которой делаем запрос
params = {
    "user_id": 1,
    "v": 5.95,
    "fields": "sex,bdate",
    "access_token": token,
    "lang": "ru",
}  # Перечисляем параметры нашего запроса в словаре params
response = requests.get(url, params=params)  # Отправляем запрос
print(response.text)  # Выводим текст ответа на экран

{"response":[{"id":1,"bdate":"10.10.1984","sex":2,"first_name":"Павел","last_name":"Дуров","can_access_closed":true,"is_closed":false}]}


Мы получили строку в JSON-формате, которую можно преобразовать в словарь с помощью метода json(), после чего можно с лёгкостью обращаться к различным полям.

Словари нагляднее выводить с помощью функции pprint(), которую мы уже использовали ранее:

In [7]:
from pprint import pprint

pprint(response.json())  # Выводим содержимое словаря, содержащего ответ, на экран

{'response': [{'bdate': '10.10.1984',
               'can_access_closed': True,
               'first_name': 'Павел',
               'id': 1,
               'is_closed': False,
               'last_name': 'Дуров',
               'sex': 2}]}


Как вы видите, по ключу response мы можем получить список, в котором хранятся словари, содержащие информацию о запрошенных нами пользователях. Мы запросили информацию лишь об одном из них, поэтому список содержит только один элемент. Извлечём его:

In [8]:
user = response.json()["response"][
    0
]  # Извлекаем из словаря по ключу response информацию о первом пользователе
print(user["bdate"])  # Выводим дату рождения первого пользователя на экран

10.10.1984


### Задание 6.1
Мы вывели на экран дату рождения пользователя. Напишите строку кода, которая выведет на экран имя пользователя:

In [9]:
print(user['first_name'])

Павел


Метод users.get() позволяет запрашивать информацию о множестве (до 1 000) пользователей одновременно. Для этого нужно использовать параметр user_ids и передавать id через запятую в строковом формате. Например, чтобы получить информацию о пользователях с id=1, id=2, id=3, необходимо передать значение параметра user_ids='1,2,3'.

Попробуем это сделать:

In [10]:
ids = ",".join(
    map(str, range(1, 4))
)  # Формируем строку, содержащую информацию о поле id первых трёх пользователей
params = {
    "user_ids": ids,
    "v": 5.95,
    "fields": "bdate",
    "access_token": token,
    "lang": "ru",
}  # Формируем строку параметров
pprint(
    requests.get(url, params=params).json()
)  # Посылаем запрос, полученный ответ в формате JSON-строки преобразуем в словарь и выводим на экран его содержимое, используя функцию pprint()

{'response': [{'bdate': '10.10.1984',
               'can_access_closed': True,
               'first_name': 'Павел',
               'id': 1,
               'is_closed': False,
               'last_name': 'Дуров'},
              {'bdate': '14.2',
               'can_access_closed': False,
               'first_name': 'Александра',
               'id': 2,
               'is_closed': True,
               'last_name': 'Владимирова'},
              {'can_access_closed': True,
               'deactivated': 'deleted',
               'first_name': 'DELETED',
               'id': 3,
               'is_closed': False,
               'last_name': ''}]}


### Задание 6.2
Используя API, определите долю женщин (sex=1) среди пользователей с id от 1 до 500. Иногда будут попадаться пользователи, у которых пол не указан (sex=0), — таких пользователей не нужно учитывать в общем числе.

В ответе укажите число, округлив до двух знаков после точки-разделителя, например, 0.55.

Пример: если у нас будет 300 пользователей с sex=1, 100 пользователей с sex=2 и 100 пользователей с sex=0, то в ответе должно быть 0.75.

In [15]:
ids = ",".join(
    map(str, range(1, 501))
)  # Формируем строку, содержащую информацию о поле id первых трёх пользователей
params = {
    "user_ids": ids,
    "v": 5.95,
    "fields": "sex, bdate",
    "access_token": token,
    "lang": "ru",
}  # Формируем строку параметров

In [16]:
response = requests.get(url, params=params).json()["response"]

In [17]:
response

[{'id': 1, 'bdate': '10.10.1984', 'sex': 2, 'first_name': 'Павел', 'last_name': 'Дуров', 'can_access_closed': True, 'is_closed': False}, {'id': 2, 'bdate': '14.2', 'sex': 1, 'first_name': 'Александра', 'last_name': 'Владимирова', 'can_access_closed': False, 'is_closed': True}, {'id': 3, 'sex': 0, 'deactivated': 'deleted', 'first_name': 'DELETED', 'last_name': '', 'can_access_closed': True, 'is_closed': False}, {'id': 4, 'sex': 0, 'deactivated': 'deleted', 'first_name': 'DELETED', 'last_name': '', 'can_access_closed': True, 'is_closed': False}, {'id': 5, 'bdate': '18.11', 'sex': 2, 'first_name': 'Илья', 'last_name': 'Перекопский', 'can_access_closed': True, 'is_closed': False}, {'id': 6, 'sex': 2, 'first_name': 'Николай', 'last_name': 'Дуров', 'can_access_closed': True, 'is_closed': False}, {'id': 7, 'sex': 2, 'first_name': 'Алексей', 'last_name': 'Кобылянский', 'can_access_closed': False, 'is_closed': True}, {'id': 8, 'sex': 2, 'first_name': 'Аки', 'last_name': 'Сепиашвили', 'can_acces

In [18]:
men = women = 0
for elem in response:
    if elem["sex"] == 2:
        men += 1
    elif elem["sex"] == 1:
        women += 1
    else:
        continue
print(round(women / (men + women), 2))

0.49


### Сбор информации из групп

Теперь мы научимся считать произвольные метрики групп, собирая данные из API и работая с двумя ограничениями, которые свойственны практически всем системам:

- ограничение на количество вызовов в единицу времени;
- ограничение на количество выгружаемых строк за один запрос.

Ограничение на количество запросов в единицу времени сделано для того, чтобы избежать чрезмерной нагрузки на серверы системы. В ряде случаев небольшое количество отчётов можно выгрузить и уложиться в действующий лимит.

Однако второе ограничение не удастся обойти в случае выгрузки больших отчётов. Например, чтобы получить список всех пользователей популярной группы, серверу, возможно, придётся отправить ответ, содержащий миллионы записей.

Давайте рассмотрим, как работать с этими ограничениями на примере выгрузки списка пользователей группы https://vk.com/vk социальной сети ВКонтакте.

Обратимся к документации, чтобы узнать, какие методы нам доступны для групп, — для получения списка пользователей группы доступен метод groups.getMembers.

Согласно документации, обязательным параметром данного метода является group_id — идентификатор, или короткое имя, группы. В нашем случае это vk: https://vk.com/vk. Протестируем, как работает метод в самом простом случае, — получим id участников группы:

In [21]:
url = "https://api.vk.com/method/groups.getMembers"  # Указываем адрес обращения
params = {
    "group_id": "vk",
    "v": 5.95,
    "access_token": token,
}  # Формируем строку параметров
response = requests.get(url, params=params)  # Посылаем запрос
data = response.json()  # Ответ сохраняем в переменной data в формате словаря
print(data)  # Выводим содержимое переменной data на экран (отображён фрагмент)

{'response': {'count': 13701658, 'items': [6, 19, 47, 54, 79, 177, 198, 212, 219, 243, 296, 345, 353, 421, 431, 450, 467, 485, 510, 550, 619, 640, 690, 702, 721, 804, 809, 832, 834, 847, 900, 905, 907, 914, 930, 943, 952, 958, 966, 976, 997, 1000, 1018, 1023, 1032, 1033, 1039, 1059, 1097, 1131, 1139, 1140, 1159, 1174, 1185, 1188, 1273, 1301, 1333, 1334, 1351, 1381, 1386, 1388, 1406, 1411, 1418, 1432, 1494, 1503, 1531, 1550, 1568, 1586, 1590, 1593, 1598, 1610, 1615, 1632, 1634, 1650, 1679, 1690, 1697, 1698, 1699, 1700, 1721, 1740, 1754, 1796, 1814, 1820, 1829, 1834, 1839, 1840, 1843, 1858, 1863, 1869, 1887, 1889, 1917, 1943, 1947, 1955, 1969, 2019, 2028, 2050, 2051, 2052, 2059, 2077, 2103, 2136, 2150, 2195, 2201, 2230, 2236, 2273, 2281, 2296, 2298, 2376, 2389, 2395, 2403, 2412, 2436, 2456, 2466, 2470, 2484, 2513, 2515, 2539, 2576, 2592, 2601, 2622, 2644, 2654, 2692, 2706, 2745, 2755, 2767, 2787, 2797, 2827, 2858, 2909, 2919, 2922, 2932, 2953, 2962, 2982, 3001, 3006, 3041, 3130, 3135, 31

По ключу count мы можем получить общее число участников группы, а список по ключу items хранит их id. Посмотрим на него поближе:

In [22]:
print(len(data["response"]["items"]))  # Выводим на экран количество элементов словаря

1000


Мы видим, что всего пользователей в группе больше 11 миллионов, а получили мы только первую тысячу пользователей группы. По информации, указанной в документации о параметре count, это максимум, который может отдать API за один раз.

Для получения следующей тысячи пользователей можно воспользоваться параметром offset (с англ. смещение), который передвинет начало отсчёта. Для выгрузки всех пользователей группы будем в цикле выгружать по 1000 пользователей (count будет всегда равен 1000), увеличивая смещение offset на величину count.

Для тренировки напишем цикл выгрузки первых 20 пользователей со значением count=5. Иными словами, мы будем выгружать по пять пользователей за запрос до тех пор, пока не получим информацию о 20 пользователях.

Давайте выведем на экран первые 20 пользователей из нашей первой попытки получить информацию о 1000 пользователей, чтобы мы могли сверить результат выгрузки из 20 пользователей:

In [23]:
users_for_checking = data["response"]["items"][
    :20
]  # Загружаем в переменную информацию об id первых 20 пользователей в виде списка
print(users_for_checking)  # Выводим перечень id первых 20 пользователей

[6, 19, 47, 54, 79, 177, 198, 212, 219, 243, 296, 345, 353, 421, 431, 450, 467, 485, 510, 550]


Теперь используем count и offset, чтобы получить те же id по пять за раз:

In [24]:
url = "https://api.vk.com/method/groups.getMembers"  # Указываем адрес обращения
count = 5
offset = 0
user_ids = []
max_count = 20
while offset < max_count:
    # Будем выгружать по count=5 пользователей,
    # начиная с того места, где закончили на предыдущей итерации (offset)
    print("Выгружаю {} пользователей с offset = {}".format(count, offset))
    params = {
        "group_id": "vk",
        "v": 5.95,
        "count": count,
        "offset": offset,
        "access_token": token,
    }
    response = requests.get(url, params=params)
    data = response.json()
    user_ids += data["response"]["items"]
    # Увеличиваем смещение на количество строк, которое мы уже выгрузили
    offset += count
print(user_ids)

Выгружаю 5 пользователей с offset = 0
Выгружаю 5 пользователей с offset = 5
Выгружаю 5 пользователей с offset = 10
Выгружаю 5 пользователей с offset = 15
[6, 19, 47, 54, 79, 177, 198, 212, 219, 243, 296, 345, 353, 421, 431, 450, 467, 485, 510, 550]


Сравним списки, полученные двумя способами:

In [25]:
print(user_ids == users_for_checking)

True


### Ограничение по частоте запросов

В API часто добавляют ограничение по частоте запросов, чтобы отдельно взятые пользователи слишком сильно не перегружали сервер. Подобное ограничение есть и у ВКонтакте — в документации указано, что можно делать не более трёх запросов в секунду.

Чтобы не следить за частотой отправки запросов с секундомером в руках, мы можем после каждого запроса делать паузу. В этом случае, даже если код будет выполняться на самом быстром компьютере, мы не нарушим установленное ограничение, так как периодичность отправки запросов будет искусственно замедлена.

Воспользуемся библиотекой time и методом sleep, с помощью которого мы можем добавить паузу, например в 0.5 секунд, после каждого запроса:

In [27]:
import time

In [28]:
url = "https://api.vk.com/method/groups.getMembers"  # Указываем адрес страницы, к которой делаем запрос
count = 1000
offset = 0
user_ids = []
while offset < 5000:
    params = {
        "group_id": "vk",
        "v": 5.95,
        "count": count,
        "offset": offset,
        "access_token": token,
    }
    response = requests.get(url, params=params)
    data = response.json()
    user_ids += data["response"]["items"]
    offset += count
    print("Ожидаю 0.5 секунды...")
    time.sleep(0.5)
print("Цикл завершен, offset =", offset)

Ожидаю 0.5 секунды...
Ожидаю 0.5 секунды...
Ожидаю 0.5 секунды...
Ожидаю 0.5 секунды...
Ожидаю 0.5 секунды...
Цикл завершен, offset = 5000


### Лайки, репосты и комментарии

Через API новостной ленты ВКонтакте мы можем получить информацию о взаимодействии с сообщениями в ленте.

Для примера продолжим работать с группой https://vk.com/vk и рассмотрим последние 100 сообщений в новостной ленте.

Для получения информации о сообщениях на стене в API ВКонтакте предусмотрен метод wall.get. Применим его:

In [29]:
url = "https://api.vk.com/method/wall.get"  # Указываем адрес страницы, к которой делаем запрос
params = {
    "domain": "vk",
    "filter": "owner",
    "count": 1000,
    "offset": 0,
    "access_token": token,
    "v": 5.95,
}
response = requests.get(url, params=params)
pprint(response.json())

{'response': {'count': 775,
              'items': [{'attachments': [{'photo': {'access_key': 'e7c80cdbe1f2054e51',
                                                    'album_id': -7,
                                                    'date': 1732184999,
                                                    'id': 457354500,
                                                    'orig_photo': {'height': 1177,
                                                                   'type': 'base',
                                                                   'url': 'https://sun9-36.userapi.com/s/v1/ig2/x-afxfZaFGoc9TpyG0eiYrZNWanW3cBkL013Ug8b4XYJPxSq0ghrey88NjliIsPypx483q5Q10e1M1zbGF7LQFZp.jpg?quality=96&as=32x29,48x44,72x66,108x99,160x147,240x221,360x331,480x441,540x497,640x588,720x662,1080x993,1280x1177&from=bu',
                                                                   'width': 1280},
                                                    'owner_id': -22822305,
                      

Посмотрим на количество результатов:

In [30]:
len(response.json()["response"]["items"])
## 100

100

Посмотрим на информацию об отдельном сообщении:

In [ ]:
response.json()["response"]["items"][0]

{'inner_type': 'wall_wallpost', 'comments': {'count': 1821}, 'marked_as_ads': 0, 'hash': 'DY2LgnT2fzk7eQgGrQ', 'type': 'post', 'push_subscription': {'is_subscribed': False}, 'attachments': [{'type': 'photo', 'photo': {'album_id': -7, 'date': 1732184999, 'id': 457354500, 'owner_id': -22822305, 'access_key': 'e7c80cdbe1f2054e51', 'post_id': 1527823, 'sizes': [{'height': 120, 'type': 'm', 'width': 130, 'url': 'https://sun9-36.userapi.com/s/v1/ig2/x-afxfZaFGoc9TpyG0eiYrZNWanW3cBkL013Ug8b4XYJPxSq0ghrey88NjliIsPypx483q5Q10e1M1zbGF7LQFZp.jpg?quality=96&as=32x29,48x44,72x66,108x99,160x147,240x221,360x331,480x441,540x497,640x588,720x662,1080x993,1280x1177&from=bu&cs=130x120'}, {'height': 120, 'type': 'o', 'width': 130, 'url': 'https://sun9-36.userapi.com/s/v1/ig2/x-afxfZaFGoc9TpyG0eiYrZNWanW3cBkL013Ug8b4XYJPxSq0ghrey88NjliIsPypx483q5Q10e1M1zbGF7LQFZp.jpg?quality=96&as=32x29,48x44,72x66,108x99,160x147,240x221,360x331,480x441,540x497,640x588,720x662,1080x993,1280x1177&from=bu&cs=130x120'}, {'heig

В полях comments, likes и reposts содержится статистика по взаимодействию с сообщением пользователей (на момент получения информации) — число комментариев, лайков и репостов.

Давайте соберём итоговую статистику для последних десяти непустых сообщений в словарь stats. В качестве ключа будем использовать начало сообщения (если начало сообщения пустое, то информацию о таком сообщении проигнорируем), в качестве значения — список с тремя интересующими нас метриками и временем публикации (комментарии, лайки, репосты, дата публикации):

In [32]:
stats = {}
count_post = 0  # Счётчик «непустых» сообщений
for record in response.json()["response"]["items"][:]:
    title = record["text"][:30]
    if title:
        stats[title] = [
            record["comments"]["count"],
            record["likes"]["count"],
            record["reposts"]["count"],
            record["date"],
        ]
        count_post += 1
    if count_post < 10:
        continue
    else:
        break
pprint(stats)

{'VK Музыка обновилась! Что ново': [495, 838, 245, 1728310195],
 'В чатах ВКонтакте появилась па': [1566, 485, 145, 1731411000],
 'Итоги ВКонтакте за третий квар': [1821, 433, 102, 1732186801],
 'Мы изучили портрет пользовател': [913, 332, 75, 1732002300],
 'Новая программа монетизации от': [823, 1071, 362, 1727083149],
 'Обновился VK Мессенджер в веб-': [494, 837, 341, 1726748100],
 'Представляем VK AdBlogger — но': [162, 468, 181, 1726654200],
 'Сегодня ВКонтакте исполнилось ': [647, 1785, 275, 1728560919],
 'Спешим поделиться крайне милым': [2267, 3503, 699, 1730375700],
 'Что можно успеть за 60 секунд?': [436, 1782, 195, 1726757657]}


## 7. Как настроить регулярную выгрузку данных

### Как настроить автоматический запуск

Во всех операционных системах есть встроенные средства запуска программ по расписанию. Однако можно задать расписание запуска необходимых вам функций на языке Python внутри скрипта.

Исполняемый по расписанию код часто называют задачей (англ. task). Для планирования задач в Python есть несколько библиотек, среди которых — популярный и простой в использовании модуль schedule (c англ. расписание). Он позволяет запускать код как с определённым интервалом, так и в заданное время.

Модуль schedule не входит в стандартную библиотеку Python, поэтому его необходимо установить:

In [1]:
import schedule # Импортируем модуль schedule

### Постановка задачи

Рассмотрим вариант автоматического запуска простой функции, которая выводит на экран короткое сообщение:

In [2]:
def task():
    print("Hello! I am a task!")
    return

Предположим, что мы хотим, чтобы функция запускалась через определённые интервалы времени, а именно каждые 15 минут.

Для запуска задачи через определённые интервалы времени в модуле schedule используется метод every(), который получает в качестве единственного аргумента число, указывающее, как часто следует запускать код.

Далее вызывается метод, определяющий единицы измерения промежутков времени, через которые будет выполняться функция. В нашем примере это минуты. Вот как будет выглядеть итоговый код:

In [3]:
schedule.every(15).minutes.do(task)

Every 15 minutes do task() (last run: [never], next run: 2024-11-27 22:12:09)

Если бы мы хотели запускать задачу, например, каждый час, то могли бы написать:

In [4]:
schedule.every(1).hour.do(task) 

Every 1 hour do task() (last run: [never], next run: 2024-11-27 22:57:32)

### Выполнение функции

После того как мы создали нашу функцию и определились со временем её запуска, мы можем запустить наш менеджер расписания (schedule). Для этого надо создать бесконечный цикл. Да-да, именно так, ведь скрипт должен выполняться постоянно, чтобы постоянно проверять, не пришло ли время снова выполнить задачу.

Внутри цикла мы будем вызывать особый метод run_pending() для объекта schedule, который будет проверять, нет ли задачи, которую пора выполнить.

После вызова метода run_pending() нужно будет сделать небольшую паузу, после которой можно будет снова проверять, не пришло ли время для выполнения какой-либо функции.

Давайте напишем этот код.

Для создания паузы мы будем использовать метод sleep из модуля time, поэтому наш код начнётся с импорта данного модуля:

In [ ]:
import time

while True:
    schedule.run_pending()
    time.sleep(1)

Всё готово — теперь можно сохранить файл и выполнить код.

Вот какой код в итоге получился:

In [ ]:
def task():
    print("Hello! I am a task!")
    return


schedule.every(15).minutes.do(task)

while True:
    schedule.run_pending()
    time.sleep(1)

Этот код будет каждую секунду проверять, не надо ли выполнить какую-то задачу, и раз в 15 минут будет выводить на экран фразу: "Hello! I am a task!" Вывод сообщения будет повторяться до тех пор, пока вы не остановите выполнение скрипта.